OneSignal/onesignal-expo-plugin
# Provide ability to specify custom iOS Notification Service Extension.

# Description
## One Line Summary
Allow for a custom Notification Service Extension to be specified for iOS in the app.json file.

## Details

By providing `nse` for the onesignal-expo-plugin plugin, an app can implement and provide a custom Notification Service Extension, described in detail [here](https://documentation.onesignal.com/docs/service-extensions#ios-notification-service-extension). These options make it easier for the app to stay in a managed workflow as no platform-specific actions are now required to provide a custom NSE for iOS.

### Motivation
Closes #133 

### Scope
This is an enhancement that makes changes to the expo prebuild process only, no runtime changes are expected. The additional parameters will copy the Notification Service Extension from a user-provided location, falling back to the existing pre-packaged NSE [here](https://github.com/OneSignal/onesignal-expo-plugin/blob/main/support/serviceExtensionFiles/NotificationService.m).


# Testing

## Manual testing
Various scenarios revolving around specifying `nse` were performed. The resulting `ios/OneSignalNotificationServiceExtension/NotificationService.m` file was analyzed to ensure it has been copied from the correct place.  The resulting build was run on an iOS emulator to the extension was appropriately built/configured.


# Checklist
## Overview
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR

## Testing

## Final pass
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.
